rfcs/0049-flakes.md

```diff
@@ -0,0 +1,709 @@
+            "inputs": {},
+            "narHash": "sha256-p7UqhvhwS5MZfqUbLbFm+nfG/SMJrgpNXxWpRMFif8c=",
+            "uri": "github:NixOS/nixpkgs/4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082"
+        }
+    }
+}
+```
+
+Thus, when we build this flake, the input `nixpkgs` is mapped to
+`github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e`. Nix
+will also check that the content hash of the input is equal to the one
+recorded in the lock file. This check is superfluous for Git
+repositories (since the commit hash serves a similar purpose), but for
+GitHub archives, we cannot directly check that the contents match the
+commit hash.
+
+Note that lock files are only used at top-level: the `flake.lock` files
+in dependencies (if they exist) are ignored.
+
+Nix automatically creates a `flake.lock` file when you build a local
+repository (e.g. `nix build /path/to/repo`). It will also update the
+lock file if inputs are added or removed. You can pass
+`--recreate-lock-file` to force Nix to recreate the lock file from
+scratch (and thus check for the latest version of each input).
+
+Lock files are not sufficient by themselves to ensure reproducible
+evaluation. It is also necessary to prevent certain impurities. In
+particular, the `nix` command now defaults to evaluating in "pure"
+mode, which means that the following are disallowed:
+
+* Access to files outside of the top-level flake or its inputs, as
+  well as paths fetched using `fetchTarball`, `fetchGit` and so on
+  with a commit hash or content hash. In particular this means that
+  Nixpkgs will not be able to use `~/.config/nixpkgs` anymore.
+
+* Access to the environment. This means that `builtins.getEnv "<var>"`
+  always returns an empty string.
+
+* Access to the system type (`builtins.currentSystem`).
+
+* Access to the current time (`builtins.currentTime`).
+
+* Use of the Nix search path (`<...>`); composition must be done
+  through flake inputs or `fetchX` builtins.
+
+Pure mode can be disabled by passing `--impure` on the command line.
+
+## Evaluation caching
+
+The fact that evaluation is now hermetic allows `nix` to cache flake
+attributes. For example (doing `nix build` on an already present
+package):
+
+    $ time nix build nixpkgs:firefox
+    real    0m1.497s
+
+    $ time nix build nixpkgs:firefox
+    real    0m0.052s
+
+The evaluation cache is kept in `~/.cache/nix/eval-cache-v1.sqlite`,
+which has entries like
+
+    INSERT INTO Attributes VALUES(
+      X'92a907d4efe933af2a46959b082cdff176aa5bfeb47a98fabd234809a67ab195',
+      'packages.firefox',
+      1,
+      '/nix/store/pbalzf8x19hckr8cwdv62rd6g0lqgc38-firefox-67.0.drv /nix/store/g6q0gx0v6xvdnizp8lrcw7c4gdkzana0-firefox-67.0 out');
+
+where the hash `92a9...` is a fingerprint over the flake store path
+and the contents of its lockfile. Because flakes are evaluated in pure
+mode, this uniquely identifies the evaluation result.
+
+Currently caching is only done for top-level attributes (e.g. for
```

Comment:
"top-level attributes" is confusing here — should this be "attributes specified in a command"? Because as far as I can tell, `packages.firefox` is _not_ a top-level attribute.